src/ray/raylet/scheduling/policy/bundle_scheduling_policy.cc

```diff
@@ -16,21 +16,62 @@
 
 namespace {
 
+/// Return true if scheduling this bundle (with resource_request) will exceed the
+/// max cpu fraction for placement groups.
+///
+/// \param node_resources The resource of the current node.
+/// \param resource_request The requested resources for the current bundle.
+/// \param max_cpu_fraction_per_node Highest CPU fraction the bundles can take up.
+/// \param available_cpus_before_scheduling Available CPUs on this node before
+///   scheduling the curernt bundles. It is used to calculate how many CPUs are
+///   allocated by the current bundles so far. It will help us figuring out
+///   the total CPU allocation from the current bundles for this node.
 bool AllocationWillExceedMaxCpuFraction(const ray::NodeResources &node_resources,
                                         const ray::ResourceRequest &resource_request,
-                                        double max_cpu_fraction_per_node) {
+                                        double max_cpu_fraction_per_node,
+                                        double available_cpus_before_scheduling) {
```

Comment:
hmm this looks a bit complicated... will it be simpler if we track how many cpus has been allocated for placement groups

Reply:
So we need to know 1. track how many pgs have been allocated per node 2. how many CPUs have been allocated in the current scheduling context per node (current bundle scheduling context).

To support `we track how many cpus has been allocated for placement groups`, we should have some data structure to track this. It might be possible, so let me actually take a look